packages/plugin-ext/src/plugin/node/debug/debug.ts

```diff
@@ -331,15 +334,60 @@ export class DebugExtImpl implements DebugExt {
         return undefined;
     }
 
+    private getConfigurationProviderRecord(handle: number): {
+        provider: theia.DebugConfigurationProvider
+        type: string
+    } {
+        const record = this.configurationProviders.find(p => p.handle === handle);
+        if (!record) {
+            throw new Error('No Debug configuration provider found with given handle number: ' + handle);
+        }
+        const { provider, type } = record;
+        return { provider, type };
+    }
+
+    async $provideDebugConfigurationsByHandle(handle: number, workspaceFolderUri: string | undefined): Promise<theia.DebugConfiguration[]> {
+        const { provider, type } = this.getConfigurationProviderRecord(handle);
+
+        const configurations = await provider.provideDebugConfigurations?.(this.toWorkspaceFolder(workspaceFolderUri));
+        if (!configurations) {
+            throw new Error('nothing returned from DebugConfigurationProvider.provideDebugConfigurations, type: ' + type);
+        }
+
+        return configurations;
+    }
+
+    async $resolveDebugConfigurationByHandle(
+        handle: number,
+        workspaceFolderUri: string | undefined,
+        debugConfiguration: theia.DebugConfiguration): Promise<theia.DebugConfiguration | undefined> {
+
+        const { provider } = this.getConfigurationProviderRecord(handle);
+        return provider.resolveDebugConfiguration?.(this.toWorkspaceFolder(workspaceFolderUri), debugConfiguration);
+    }
+
+    async $resolveDebugConfigurationWithSubstitutedVariablesByHandle(
+        handle: number,
+        workspaceFolderUri: string | undefined,
+        debugConfiguration: theia.DebugConfiguration): Promise<theia.DebugConfiguration | undefined> {
+
+        const { provider } = this.getConfigurationProviderRecord(handle);
+        return provider.resolveDebugConfigurationWithSubstitutedVariables?.(this.toWorkspaceFolder(workspaceFolderUri), debugConfiguration);
+    }
+
     async $provideDebugConfigurations(debugType: string, workspaceFolderUri: string | undefined, dynamic: boolean = false): Promise<theia.DebugConfiguration[]> {
         let result: theia.DebugConfiguration[] = [];
 
-        const providers = dynamic ? this.dynamicConfigurationProviders.get(debugType) : this.configurationProviders.get(debugType);
-        if (providers) {
-            for (const provider of providers) {
-                if (provider.provideDebugConfigurations) {
-                    result = result.concat(await provider.provideDebugConfigurations(this.toWorkspaceFolder(workspaceFolderUri)) || []);
-                }
+        const trigger = dynamic ? DebugConfigurationProviderTriggerKind.Dynamic : DebugConfigurationProviderTriggerKind.Initial;
+        const providers = this.configurationProviders.
+            filter(p =>
+                (p.type === debugType) &&
+                (p.trigger === trigger)).
+            map(p => p.provider);
```

Comment:
In JavaScript the convention is to keep the `.method(` on the same line, unlike `.<\n>method`:

```suggestion
        const providers = this.configurationProviders
            .filter(p => (
                p.type === debugType &&
                p.trigger === trigger
            ))
            .map(p => p.provider);
```